internal/btf/btf.go

```diff
@@ -484,14 +491,44 @@ func (s *Spec) Program(name string, length uint64) (*Program, error) {
 	return &Program{s, length, funcInfos, lineInfos, relos}, nil
 }
 
-// FindType searches for a type with a specific name.
+// TypeByID returns the BTF Type with the given type ID.
+//
+// Returns an error wrapping ErrNotFound if a Type with the given ID
+// does not exist in the Spec.
+func (s *Spec) TypeByID(id TypeID) (Type, error) {
+	if int(id) > len(s.types) {
+		return nil, fmt.Errorf("type ID %d: %w", id, ErrNotFound)
+	}
+	return s.types[id], nil
+}
+
+// TypesByName returns a list of BTF Types with the given name.
 //
-// Called T a type that satisfies Type, typ must be a non-nil **T.
-// On success, the address of the found type will be copied in typ.
+// Multiple Types of the same name can exist in case multiple 'struct flavors'
+// are described in BTF info. These are used to deal with changes in kernel
+// data structures.
 //
 // Returns an error wrapping ErrNotFound if no matching
-// type exists in spec.
-func (s *Spec) FindType(name string, typ interface{}) error {
+// Type exists in the Spec.
+func (s *Spec) TypesByName(name string) ([]Type, error) {
```

Comment:
It's possible for multiple types `foo` to exist even without flavours since BTF is merged from multiple compilation units without name spacing. Put another way, the `[]Type` returned from this function may contain types with identical names. It's also possible for multiple types with the same flavour name to exist, but lets not go there. TypeByName flags this problem by returning an error if there are ambiguous matches, and allows resolving it by passing a name including a flavour.

1. TypeByName("foo___bar") returns the exact flavour.
2. TypeByName("foo") returns the single type that has the name "foo".

Case 2 is somewhat problematic since it doesn't find `foo___bar` if that is the only existing type.

Your proposed TypesByName instead returns all types with the same essential name, so `TypesByName("foo___bar")` will return types named `foo`. How about the following instead:

* TypesByName("foo___bar") returns all types for the flavour
* TypesByName("foo") returns all types with essentialName == foo
* TypeByName("foo") returns a type with essentialName == foo as long as there is only a single one

That last point may make some lookups fail that currently work, but it's unlikely to cause issues.

Another discrepancy is that this function doesn't have a `typ` parameter like `TypeByName` does. I'm working on generating types from bpf2go, and for that I wrote [`FindAnyType`](https://github.com/lmb/ebpf/blob/820745cec7d5822f83ed15441c48a26bf54662c5/internal/btf/btf.go#L526):

```Go
func (s *Spec) FindAnyType(name string) (NamedType, error)
```

How about we call the methods `AnyType(s)ByName` and `Type(s)ByName`, with `TypesByName` being unimplemented since it doesn't seem super useful?

Reply:
That all sounds good :+1: I'm just starting out learning about struct flavors, so not aware of many of the corner cases yet. Could we add tests for the gotchas you've outlined?

My main goal with this proposal was to give the caller the flexibility to resolve ambiguity on their own, by their own rules, to avoid having to overload the lib with specialized filtering logic. I think it makes sense to have this as a general building block.